Prevent analytics collector from missing events when tower is under stress

jladdjr opened the pull request:
Adapts https://github.com/ansible/awx/pull/9358#issuecomment-797747525 to work in the context of awx using partitioned job events tables (see https://github.com/ansible/awx/issues/9039 and https://github.com/ansible/awx/pull/9341).

Commit:
whitespace clean-up

Commit:
collect job events based on job event modified time

* when collecting job events by creation time
  it is possible to miss events that were created
  at one point, but actually committed to the db
  much later.
* since events' modified time is set when they are
  committed to the db, we shouldn't miss any job events
* selecting job events by modified time wasn't possible
  beforehand because we didn't have an index for
  jobevent's modified field

Commit:
query for jobevents based on table location

* pre-migration jobevents live in unpartitioned table
  where only created field has index
* post-migration jobevents live in partitions
  where modified field has index
  (and should be used to ensure no events are missing)

Commit:
collect jobevent analytics by partition when collecting up to hrzn

* only applies if all job events going back to horizon
  are stored in partitioned tables

Commit:
add missing call to partition_within_horizon

jladdjr commented:
cc @chrismeyersfsu @Ladas @jbradberry @kdelee

jbradberry commented on awx/main/analytics/collectors.py:
```diff
@@ -4,6 +4,8 @@
 import os.path
 import platform
 import distro
+import pytz
```
You should probably use the `utc` from django.utils.timezone instead of importing pytz.

jbradberry commented on awx/main/analytics/collectors.py:
```diff
@@ -78,7 +80,51 @@ def events_slicing(key, since, until):
+    # if we are collecting job events starting at the horizon
+    # and all job events are stored in partitions,
+    # slice by partition
+    if lower == horizon and lower > partition_epoch:
```
I don't agree with limiting this optimization to the case where the lower bound is exactly at the 4-week limit.

jladdjr replied:
agreed, rewriting this 👍

Commit:
send JobEvent.job_created to analytics

Commit:
create custom key to compare jobevent `lastentry`

* as part of the move to partition-based job event tables
  we will need to migrate how the last entry of a job event was
  recorded, moving from a pk to partition-table based entry

Commit:
collect job events from both unpartitioned and partitioned tables

Commit:
WIP unit tests

jladdjr commented on awx/main/analytics/collectors.py:
```diff
@@ -78,16 +104,26 @@ def events_slicing(key, since, until):
-    if not since and last_entries.get(key):
-        previous_pk = max(last_entries[key], previous_pk)
-    final_pk = pk_values['pk__max'] or 0
+    # Stage 1: Process any job events in unpartitioned job event table
```
@jbradberry this is still a work in progress. We still need to get some data on the (1) only collect data by scooping up partitions approach vs the (2) get partitions when going up to the horizon, otherwise collect by pk / modified field.

This is written to follow the second approach, but I'd actually like to have it so that we can toggle between both options to make testing at scale easier.

jladdjr commented on awx/main/analytics/core.py:
```diff
@@ -52,7 +52,7 @@ def all_collectors():
 
 
-def register(key, version, description=None, format='json', expensive=None):
+def register(key, version, description=None, format='json', expensive=None, compare_key=None):
```
@jbradberry I'm extending this a bit so that, when needed, we can customize how we compare entries. The key function for job events is [here](https://github.com/ansible/awx/pull/9856/files#diff-5e7230fef85166a18967e2216535cd734d66de8b26983c303981e3c82b6fa1afR383).

jladdjr commented on awx/main/tests/functional/analytics/test_collectors.py:
```diff
@@ -72,6 +73,29 @@ def write_stdout(self, sql, fd):
     return path
 
 
+@pytest.mark.django_db
```
I'm still working on the unit tests, but want to get several of these in so that we can make sure we're collecting job events the way we expect.

jladdjr commented on awx/main/analytics/collectors.py:
```diff
@@ -344,39 +380,56 @@ def _copy_table(table, query, path):
 
 
-@register('events_table', '1.2', format='csv', description=_('Automation task records'), expensive=events_slicing)
+def jobevent_lastentry_cmp(lastentry):
```
When `analytics/core.py` uses `max` to compare last entries for job events, we pass this key function in to ensure that pk-based entries come before partition table entries.

Commit:
collect jobevents by pk, unless collecting up to horizon

Commit:
fix unit test

Commit:
fix event query

Commit:
unit test for gathering old / new job event records

Commit:
add test_collect_job_events_from_partitioned_table

Commit:
add test_collect_job_events_up_through_horizon

jbradberry commented on awx/main/analytics/core.py:
```diff
@@ -243,7 +246,7 @@ def gather(dest=None, module=None, subset=None, since=None, until=None, collecti
                         if collection_type != 'dry-run':
                             with disable_activity_stream():
-                                last_entries[key] = max(last_entries[key], end) if last_entries.get(key) else end
+                                last_entries[key] = max(last_entries[key], end, key=compare_key) if last_entries.get(key) else end
```
The support for `key=None` was only added in Python 3.8.

jbradberry reviewed (changes requested):
I don't see anything here that addresses the concern of what should happen with dirty partitions that have been marked as already shipped.
